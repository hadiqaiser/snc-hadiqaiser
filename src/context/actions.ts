export enum UserAction {
  SetEnableLogs = "SET_ENABLE_LOGS",
  SetSelectedUser = "SET_SELECTED_USER",
  SetUserData = "SET_USER_DATA",
}
