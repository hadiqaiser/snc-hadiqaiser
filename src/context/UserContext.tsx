import { createContext, ReactNode, useContext, useReducer } from "react";

// Actions
import { UserAction } from "./actions";

// Types
import { Person, User } from "@/utils/common/person";

type UserState = {
  enableLogs: boolean;
  selectedUser: Person | undefined;
  userData: User;
};

type UserContextValue = {
  state: UserState;
  dispatch: React.Dispatch<Action>;
};

type Action =
  | { type: UserAction.SetEnableLogs; payload: boolean }
  | { type: UserAction.SetSelectedUser; payload: Person | undefined }
  | { type: UserAction.SetUserData; payload: User };

const defaultValues: UserState = {
  enableLogs: false,
  selectedUser: undefined,
  userData: {
    backgroundImageUrl: "",
    profilePictureUrl: "",
    name: "",
    title: "",
    followers: 0,
    following: 0,
  },
};

interface Props {
  children: ReactNode;
}

export const UserContext = createContext<UserContextValue>({
  state: defaultValues,
  dispatch: () => null,
});

export const UserReducer = (state: UserState, action: Action): UserState => {
  switch (action.type) {
    case UserAction.SetEnableLogs:
      return { ...state, enableLogs: action.payload };
    case UserAction.SetSelectedUser:
      return { ...state, selectedUser: action.payload };
    case UserAction.SetUserData:
      return { ...state, userData: action.payload };
    default:
      return state;
  }
};

export const UserProvider: React.FC<Props> = ({ children }: Props) => {
  const [state, dispatch] = useReducer(UserReducer, defaultValues);

  return (
    <UserContext.Provider value={{ state, dispatch }}>
      {children}
    </UserContext.Provider>
  );
};

export const useUserContext = (): UserContextValue => {
  const safeContext = useContext(UserContext);
  if (!safeContext) {
    throw new Error("useUserContext must be used within a UserProvider");
  }

  return safeContext;
};
