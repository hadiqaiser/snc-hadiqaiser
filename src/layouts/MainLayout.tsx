import {
  useCallback,
  useEffect,
  FunctionComponent,
  PropsWithChildren,
  useState,
} from "react";
import classNames from "classnames";
import axios from "axios";
import { QueryFunctionContext, useQuery } from "@tanstack/react-query";
import { toast } from "sonner";

// Custom Component
import { DisplayCurrentTime } from "@/views/home/DisplayCurrentTime";
import { LogSwitch } from "@/views/home/LogSwitch";
import { UserErrorAlert } from "@/views/home/UserErrorAlert";
import { UserInfoCard } from "@/views/home/UserInfoCard";
import { UserSelector } from "@/views/home/UserSelector";

// Contexts
import { useUserContext } from "@/context/UserContext";
import { UserAction } from "@/context/actions";

// Hooks
import { useLogUserDetail } from "@/hooks/useLogUserDetail";

// Types
import { Person, User } from "@/utils/common/person";

// Font
import { Inter } from "next/font/google";

const inter = Inter({ subsets: ["latin"] });

type MainLayoutProps = {};

export const MainLayout: FunctionComponent<
  PropsWithChildren<MainLayoutProps>
> = () => {
  // State
  const [currentTime, setCurrentTime] = useState<Date>(new Date());

  // Hooks
  const {
    state: { selectedUser },
    dispatch,
  } = useUserContext();

  useLogUserDetail(currentTime);

  const updateCurrentTime = useCallback((time: Date) => {
    setCurrentTime(time);
  }, []);

  const getUserData = useCallback(
    async ({
      queryKey,
      signal,
    }: QueryFunctionContext<[string, Person]>): Promise<User> => {
      const [, selectedUser] = queryKey;
      const response = await axios.get(`/api/person?person=${selectedUser}`, {
        signal,
      });

      return response.data;
    },
    [],
  );

  const { data, error, isSuccess, isFetching } = useQuery({
    queryKey: ["users", selectedUser as Person],
    queryFn: getUserData,
    enabled: !!selectedUser,
  });

  useEffect(() => {
    if (error) {
      toast.error(`Failed to fetch ${selectedUser} data!`);
    }
  }, [dispatch, error, selectedUser]);

  useEffect(() => {
    if (isSuccess && !isFetching) {
      dispatch({ type: UserAction.SetUserData, payload: data as User });
      toast.success(`Successfully fetched ${selectedUser} data!`);
    }
  }, [data, dispatch, isFetching, isSuccess, selectedUser]);

  return (
    <main
      className={classNames(
        inter.className,
        "h-screen w-screen",
        "flex flex-col p-6 items-center justify-between",
      )}
    >
      <section className={classNames("flex flex-col items-center w-full")}>
        {error ? <UserErrorAlert /> : <UserInfoCard loading={isFetching} />}
        <UserSelector />
      </section>
      <section
        className={classNames(
          "flex flex-row justify-between w-full sm:w-8/12 border border-gray-300 rounded-lg p-4",
        )}
      >
        <DisplayCurrentTime
          currentTime={currentTime}
          updateCurrentTime={updateCurrentTime}
        />
        <LogSwitch />
      </section>
    </main>
  );
};
