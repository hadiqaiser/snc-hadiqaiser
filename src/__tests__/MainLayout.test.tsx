import { render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import { useQuery } from "@tanstack/react-query";
import { toast } from "sonner";

// Layout
import { MainLayout } from "@/layouts/MainLayout";

// Context
import { useUserContext } from "@/context/UserContext";
import { UserAction } from "@/context/actions";

// Mock dependencies
jest.mock("../context/UserContext", () => ({
  useUserContext: jest.fn(),
}));

jest.mock("@tanstack/react-query", () => ({
  useQuery: jest.fn(),
}));

jest.mock("axios");

jest.mock("sonner", () => ({
  toast: {
    error: jest.fn(),
    success: jest.fn(),
  },
}));

jest.mock("../hooks/useLogUserDetail", () => ({
  useLogUserDetail: jest.fn(),
}));

describe("MainLayout", () => {
  const mockDispatch = jest.fn();
  const mockUserData = {
    backgroundImageUrl: "/images/fallbackBgImg.svg",
    profilePictureUrl: "/images/fallbackProfileImg.svg",
    name: "John Doe",
    title: "Software Engineer",
    followers: 100,
    following: 200,
  };
  beforeAll(() => {
    const mockDate = new Date(2024, 0, 1, 12, 0, 0); // January 1, 2024, 12:00:00
    jest.spyOn(global, "Date").mockImplementation(() => mockDate);
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  beforeEach(() => {
    jest.clearAllMocks();
    (useUserContext as jest.Mock).mockReturnValue({
      state: { selectedUser: "JohnDoe", userData: mockUserData },
      dispatch: mockDispatch,
    });

    (useQuery as jest.Mock).mockReturnValue({
      data: mockUserData,
      error: false,
      isSuccess: true,
      isFetching: false,
    });
  });

  it("should match the MainLayout snapshot", () => {
    const testRender = render(<MainLayout />);

    expect(testRender).toMatchSnapshot();
  });

  it("should render all components correctly", () => {
    render(<MainLayout />);

    expect(screen.getByText("Enable Logs")).toBeInTheDocument();
    expect(screen.getByText("John Doe")).toBeInTheDocument();
    expect(screen.getByRole("img", { name: "bg-pic" })).toHaveAttribute(
      "src",
      expect.stringContaining("/images/fallbackBgImg.svg"),
    );
    expect(screen.getByRole("img", { name: "profile-pic" })).toHaveAttribute(
      "src",
      expect.stringContaining("/images/fallbackProfileImg.svg"),
    );
  });

  it("should dispatch action and show success toast on successful data fetch", async () => {
    render(<MainLayout />);

    await waitFor(() => {
      expect(mockDispatch).toHaveBeenCalledWith({
        type: UserAction.SetUserData,
        payload: mockUserData,
      });
    });

    expect(toast.success).toHaveBeenCalledWith(
      "Successfully fetched JohnDoe data!",
    );
  });

  it("should show error toast on data fetch failure", async () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: null,
      error: true,
      isSuccess: false,
      isFetching: false,
    });

    render(<MainLayout />);

    await waitFor(() => {
      expect(toast.error).toHaveBeenCalledWith("Failed to fetch JohnDoe data!");
    });

    expect(screen.getByText("Enable Logs")).toBeInTheDocument();
    expect(screen.queryByText("John Doe")).not.toBeInTheDocument();
  });

  it("should render UserErrorAlert when there is an error", () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: null,
      error: true,
      isSuccess: false,
      isFetching: false,
    });

    render(<MainLayout />);

    expect(screen.getByTestId("alert-description")).toBeInTheDocument();
  });

  it("should render UserInfoCard with loading state when data is fetching", () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: null,
      error: false,
      isSuccess: false,
      isFetching: true,
    });

    render(<MainLayout />);

    const skeletons = screen.getAllByTestId("card-skeleton");
    expect(skeletons.length).toBeGreaterThan(0);
  });
});
