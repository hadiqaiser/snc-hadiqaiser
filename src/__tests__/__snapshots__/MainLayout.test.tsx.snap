// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainLayout should match the MainLayout snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="className h-screen w-screen overflow-auto flex flex-col p-6 items-center justify-between gap-4"
      >
        <section
          class="flex flex-col items-center w-full"
        >
          <div
            class="rounded-lg border bg-card text-card-foreground shadow-sm relative w-full sm:w-8/12 h-auto"
          >
            <div
              class="relative w-full h-52"
            >
              <img
                alt="bg-pic"
                class="rounded-md object-cover"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                src="/images/fallbackBgImg.svg"
                style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
              />
            </div>
            <div
              class="pb-6 px-6"
            >
              <div
                class="relative -mt-16 px-4 flex justify-start sm:-mt-20"
              >
                <div
                  class="relative w-32 h-32 sm:w-40 sm:h-40"
                >
                  <img
                    alt="profile-pic"
                    class="rounded-full object-cover"
                    data-nimg="fill"
                    decoding="async"
                    loading="lazy"
                    src="/images/fallbackProfileImg.svg"
                    style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                  />
                </div>
              </div>
              <div
                class="flex flex-col sm:flex-row justify-between"
              >
                <div
                  class="flex flex-col gap-2 mt-3"
                >
                  <h3
                    class="scroll-m-20 text-2xl font-semibold tracking-tight"
                  >
                    John Doe
                  </h3>
                  <p
                    class="text-xl text-muted-foreground"
                  >
                    Software Engineer
                  </p>
                </div>
                <div
                  class="flex flex-row gap-5 mt-3 justify-evenly"
                >
                  <div
                    class="flex flex-col justify-between py-3 items-center"
                  >
                    <p
                      class="text-base text-muted-foreground"
                    >
                      100
                    </p>
                    <p
                      class="text-base"
                    >
                      Followers
                    </p>
                  </div>
                  <div
                    class="shrink-0 bg-border h-full w-[1px]"
                    data-orientation="vertical"
                    role="none"
                  />
                  <div
                    class="flex flex-col justify-between py-3 items-center"
                  >
                    <p
                      class="text-base text-muted-foreground"
                    >
                      200
                    </p>
                    <p
                      class="text-base"
                    >
                      Following
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="w-full sm:w-8/12 flex flex-col sm:flex-row gap-4 justify-evenly mt-8"
          >
            <button
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
            >
              PersonA
            </button>
            <button
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
            >
              PersonB
            </button>
            <button
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
            >
              PersonC
            </button>
          </div>
        </section>
        <section
          class="flex flex-col sm:flex-row justify-between w-full sm:w-8/12 border border-gray-300 rounded-lg p-4 gap-4"
        >
          <p
            class="className leading-7"
            data-testid="current-time"
          >
            01-01-2024 : 12:00:00
          </p>
          <div
            class="flex flex-row items-center space-x-2"
          >
            <button
              aria-checked="false"
              class="peer inline-flex h-6 w-11 shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=unchecked]:bg-input"
              data-state="unchecked"
              id="toggle-logs"
              role="switch"
              type="button"
              value="on"
            >
              <span
                class="pointer-events-none block h-5 w-5 rounded-full bg-background shadow-lg ring-0 transition-transform data-[state=checked]:translate-x-5 data-[state=unchecked]:translate-x-0"
                data-state="unchecked"
              />
            </button>
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              for="toggle-logs"
            >
              Enable Logs
            </label>
          </div>
        </section>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="className h-screen w-screen overflow-auto flex flex-col p-6 items-center justify-between gap-4"
    >
      <section
        class="flex flex-col items-center w-full"
      >
        <div
          class="rounded-lg border bg-card text-card-foreground shadow-sm relative w-full sm:w-8/12 h-auto"
        >
          <div
            class="relative w-full h-52"
          >
            <img
              alt="bg-pic"
              class="rounded-md object-cover"
              data-nimg="fill"
              decoding="async"
              loading="lazy"
              src="/images/fallbackBgImg.svg"
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
            />
          </div>
          <div
            class="pb-6 px-6"
          >
            <div
              class="relative -mt-16 px-4 flex justify-start sm:-mt-20"
            >
              <div
                class="relative w-32 h-32 sm:w-40 sm:h-40"
              >
                <img
                  alt="profile-pic"
                  class="rounded-full object-cover"
                  data-nimg="fill"
                  decoding="async"
                  loading="lazy"
                  src="/images/fallbackProfileImg.svg"
                  style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                />
              </div>
            </div>
            <div
              class="flex flex-col sm:flex-row justify-between"
            >
              <div
                class="flex flex-col gap-2 mt-3"
              >
                <h3
                  class="scroll-m-20 text-2xl font-semibold tracking-tight"
                >
                  John Doe
                </h3>
                <p
                  class="text-xl text-muted-foreground"
                >
                  Software Engineer
                </p>
              </div>
              <div
                class="flex flex-row gap-5 mt-3 justify-evenly"
              >
                <div
                  class="flex flex-col justify-between py-3 items-center"
                >
                  <p
                    class="text-base text-muted-foreground"
                  >
                    100
                  </p>
                  <p
                    class="text-base"
                  >
                    Followers
                  </p>
                </div>
                <div
                  class="shrink-0 bg-border h-full w-[1px]"
                  data-orientation="vertical"
                  role="none"
                />
                <div
                  class="flex flex-col justify-between py-3 items-center"
                >
                  <p
                    class="text-base text-muted-foreground"
                  >
                    200
                  </p>
                  <p
                    class="text-base"
                  >
                    Following
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="w-full sm:w-8/12 flex flex-col sm:flex-row gap-4 justify-evenly mt-8"
        >
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
          >
            PersonA
          </button>
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
          >
            PersonB
          </button>
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
          >
            PersonC
          </button>
        </div>
      </section>
      <section
        class="flex flex-col sm:flex-row justify-between w-full sm:w-8/12 border border-gray-300 rounded-lg p-4 gap-4"
      >
        <p
          class="className leading-7"
          data-testid="current-time"
        >
          01-01-2024 : 12:00:00
        </p>
        <div
          class="flex flex-row items-center space-x-2"
        >
          <button
            aria-checked="false"
            class="peer inline-flex h-6 w-11 shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=unchecked]:bg-input"
            data-state="unchecked"
            id="toggle-logs"
            role="switch"
            type="button"
            value="on"
          >
            <span
              class="pointer-events-none block h-5 w-5 rounded-full bg-background shadow-lg ring-0 transition-transform data-[state=checked]:translate-x-5 data-[state=unchecked]:translate-x-0"
              data-state="unchecked"
            />
          </button>
          <label
            class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
            for="toggle-logs"
          >
            Enable Logs
          </label>
        </div>
      </section>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
